### PR TITLE
[Connectors API] Guard cancel and update error sync job endpoints with state machine

### PR DESCRIPTION
**Pull request by timgrein**
This PR adds state transition guards through the `ConnectorSyncJobStateMachine` to the cancel sync job and update sync job error actions.

If a sync job is in status `pending` and it gets canceled it will be directly transitioned to `canceled` instead of `canceling` (this was discussed in a slack thread). Therefore we also update `canceled_at` instead of `cancelation_requested_at`, which is used to indicate an async operation.

Two additional integration tests (cancel an in progress sync job and setting an error for an in progress sync) will be added as soon as the claim sync job endpoint is implemented (both tests are tracked in the claim sync job issue).

### CODE REVIEW

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/syncjob/ConnectorSyncJobIndexService.java [hunk 1]
```diff
@@ -227,31 +229,52 @@ public void getConnectorSyncJob(String connectorSyncJobId, ActionListener<Connec
...
-        Instant cancellationRequestedAt = Instant.now();
+        try {
+            getConnectorSyncJob(connectorSyncJobId, listener.delegateFailure((getSyncJobListener, syncJobSearchResult) -> {
+                String timestampFieldToUpdate;
+                Instant updatedTimestamp = Instant.now();
+
+                ConnectorSyncStatus nextStatus;
+                ConnectorSyncStatus prevStatus = getConnectorSyncJobStatusFromSearchResult(syncJobSearchResult);
+
+                if (ConnectorSyncStatus.PENDING.equals(prevStatus)) {
```
Comment (jedrazb) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/syncjob/ConnectorSyncJobIndexService.java:240]: Any reason we just have this logic for `PENDING` and not `SUSPENDED`? I think we should also cancel suspended sync like that to replicate Kibana logic: https://github.com/elastic/kibana/blob/main/packages/kbn-search-connectors/lib/cancel_syncs.ts#L32

Comment (jedrazb) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/syncjob/ConnectorSyncJobIndexService.java:240]: Kibana seems to take the responsibility of putting any pending and suspended jobs into a canceled stated for a user-initiated sync canceling.

Comment (timgrein) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/syncjob/ConnectorSyncJobIndexService.java:240]: Thanks for pointing this out! Then I'll update the logic to also consider `SUSPENDED` and to also update `completed_at` and `cancelation_requested_at`.

Comment (timgrein) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/syncjob/ConnectorSyncJobIndexService.java]: Adjusted with [702e2f8](https://github.com/elastic/elasticsearch/pull/105722/commits/702e2f8d0f04e21ddb194f7a68cb0c1ec030bee9)

File: x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/430_connector_sync_job_cancel.yml [hunk 2]
```diff
@@ -13,7 +13,7 @@ setup:
           service_type: super-connector
 
 ---
-"Cancel a Connector Sync Job":
+"Cancel a pending Connector Sync Job - transition to canceled directly":
```
Comment (jedrazb) [inline x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/430_connector_sync_job_cancel.yml:16]: Nit: should we also have a test for suspended job?

**Review by jedrazb — APPROVED** [review]
I left a non-blocking comment about adding e2e test for a suspended->canceled transition (but I see we have unit tests anyway), otherwise it looks good to me 🚢

**Pull request merged**